sidecar/dns/server.go

```diff
@@ -0,0 +1,288 @@
+		ServiceInstances, err = s.config.DiscoveryClient.ListInstances(filters)
+	}
+	if err != nil {
+		// TODO: what Error should we return ?
+		response.SetRcode(request, dns.RcodeServerFailure)
+		return nil, fmt.Errorf("Error while reading from registry: %s", err.Error())
+	}
+	return ServiceInstances, nil
+}
+
+func (s *Server) findMatchingServices(question dns.Question, request, response *dns.Msg,
+	serviceInstances []*client.ServiceInstance) error {
+	numOfMatchingRecords := 0
+	for _, serviceInstance := range serviceInstances {
+		endPointType := serviceInstance.Endpoint.Type
+		var ip net.IP
+		var err error
+
+		switch endPointType {
+		case "tcp":
+			ip, err = validateEndPointTypeTCPAndUDP(serviceInstance.Endpoint.Value)
+
+		case "udp":
+			ip, err = validateEndPointTypeTCPAndUDP(serviceInstance.Endpoint.Value)
+
+		case "http":
+			ip, err = validateEndPointTypeHTTP(serviceInstance.Endpoint.Value)
+
+		default:
+			continue
+		}
+		if err != nil {
+			// TODO: Do I need to skip or should I print something ?
+			continue
+		}
+		numOfMatchingRecords++
+		if ip.To4() != nil {
+			record := createARecord(question.Name, ip)
+			response.Answer = append(response.Answer, record)
+		} else if ip.To16() != nil {
+			record := createAAARecord(question.Name, ip)
+			response.Answer = append(response.Answer, record)
+		}
+	}
+	if numOfMatchingRecords == 0 {
+		//Non-Existent Domain
+		response.SetRcode(request, dns.RcodeNameError)
+		return fmt.Errorf("Non-Existent Domain	 %s", question.Name)
+
+	}
+	response.SetRcode(request, dns.RcodeSuccess)
+	return nil
+
+}
+
+func validateEndPointTypeTCPAndUDP(value string) (net.IP, error) {
+	ip, _, err := net.SplitHostPort(value)
+	if err != nil {
+		return nil, err
+	}
+	parsedIP := net.ParseIP(ip)
+	if parsedIP == nil {
+		return nil, fmt.Errorf("tcp/udp ip value %s is not a valid ip format", ip)
+	}
+	return parsedIP, nil
+}
+
+func validateEndPointTypeHTTP(value string) (net.IP, error) {
+	parsedURL, err := url.Parse(value)
+	if err != nil {
+		return nil, err
+	}
+	if parsedURL.Scheme != "http" && parsedURL.Scheme != "https" {
+		return nil, fmt.Errorf("url shceme %s should be http or https", parsedURL.Scheme)
+	}
+	host := parsedURL.Host
+	ip, _, err := net.SplitHostPort(host)
+	if err != nil {
+		return nil, err
+	}
+	parsedIP := net.ParseIP(ip)
+	if parsedIP == nil {
+		return nil, fmt.Errorf("url ip value %s is not a valid ip format", ip)
+	}
+	return parsedIP, nil
+
+}
+
+func createARecord(questionName string, ip net.IP) *dns.A {
+	record := &dns.A{
+		Hdr: dns.RR_Header{
+			Name:   questionName,
+			Rrtype: dns.TypeA,
+			Class:  dns.ClassINET,
+			Ttl:    0,
+		},
+
+		A: ip,
+	}
+	return record
+}
+
+func createAAARecord(questionName string, ip net.IP) *dns.AAAA {
+	record := &dns.AAAA{
+		Hdr: dns.RR_Header{
+			Name:   questionName,
+			Rrtype: dns.TypeAAAA,
+			Class:  dns.ClassINET,
+			Ttl:    0,
+		},
+
+		AAAA: ip,
+	}
+	return record
+}
+
```

Comment:
Creating SRV records is a no brainer here. Its a simple function, where you populate the fields with the proto, IP, port and weight/priorities information. The last two can be obtained from the controller rules.